src/utils/DirectHub.cxx

```diff
@@ -0,0 +1,943 @@
+        if (msg->done_)
+        {
+            b->set_done(msg->done_->new_child());
+        }
+        // Checks if we need to wake up the flow.
+        {
+            AtomicHolder h(lock());
+            if (fd_ < 0)
+            {
+                // Catch race condition when port is already closed.
+                b->unref();
+                return;
+            }
+            pendingQueue_.insert_locked(b);
+            totalPendingSize_ += msg->buf_.size();
+            pendingTail_ = b->data();
+            if (notRunning_)
+            {
+                notRunning_ = 0;
+            }
+            else
+            {
+                // flow already running. Skip notify.
+                return;
+            }
+        }
+        notify();
+    }
+
+private:
+    /// Called on the main executor when a read error wants to cancel the write
+    /// flow. Before calling, fd_ must be -1.
+    void shutdown()
+    {
+        HASSERT(fd_ < 0);
+        {
+            AtomicHolder h(lock());
+            if (notRunning_)
+            {
+                // Queue is empty, waiting for new entries. There will be no new
+                // entries because fd_ < 0.
+                hub_->unregister_port(this, this);
+                wait_and_call(STATE(report_and_exit));
+            }
+            // Else eventually we will get to check_for_new_message() which will
+            // flush the queue, unregister the port and exit.
+        }
+    }
+
+    Action read_queue()
+    {
+        BufferType *head;
+        {
+            AtomicHolder h(lock());
+            head = static_cast<BufferType *>(pendingQueue_.next_locked().item);
+            HASSERT(head);
+            if (head->data() == pendingTail_)
+            {
+                pendingTail_ = nullptr;
+            }
+        }
+        currentHead_.reset(head);
+        nextToWrite_ = currentHead_->data()->buf_.head();
+        nextToSkip_ = currentHead_->data()->buf_.skip();
+        nextToSize_ = currentHead_->data()->buf_.size();
+        return do_write();
+    }
+
+    Action do_write()
+    {
+        if (fd_ < 0)
+        {
+            // fd closed. Drop data to the floor.
+            totalPendingSize_ -= nextToSize_;
+            return check_for_new_message();
+        }
+        uint8_t *data;
+        unsigned len;
+        nextToWrite_ = nextToWrite_->get_read_pointer(nextToSkip_, &data, &len);
+        if (len > nextToSize_)
+        {
+            len = nextToSize_;
+        }
+        nextToSkip_ = 0;
+        nextToSize_ -= len;
+        totalPendingSize_ -= len;
+        totalWritten_ += len;
+        LOG(VERBOSE, "write %u total %zu", (unsigned)len, totalWritten_);
+        return write_repeated(
+            &selectHelper_, fd_, data, len, STATE(write_done));
+    }
+
+    Action write_done()
+    {
+        if (selectHelper_.hasError_)
+        {
+            LOG(INFO, "%p: Error writing to fd %d: (%d) %s", this, fd_, errno,
```

Comment:
LOG_ERROR ?

Reply:
I don't think so. A socket error is a normal event when a remote client disconnects.